superset/security/manager.py

```diff
@@ -860,45 +834,93 @@ def set_perm(
                 )
 
-    def assert_datasource_permission(self, datasource: "BaseDatasource") -> None:
+    def raise_for_access(
+        self,
+        database: Optional["Database"] = None,
+        datasource: Optional["BaseDatasource"] = None,
+        query: Optional["Query"] = None,
+        query_context: Optional["QueryContext"] = None,
+        table: Optional["Table"] = None,
+        viz: Optional["BaseViz"] = None,
+    ) -> None:
         """
-        Assert the the user has permission to access the Superset datasource.
+        Raise an exception if the user cannot access the resource.
 
+        :param database: The Superset database
         :param datasource: The Superset datasource
-        :raises SupersetSecurityException: If the user does not have permission
+        :param query: The SQL Lab query
+        :param query_context: The query context
+        :param table: The Superset table (requires database)
+        :param viz: The visualization
+        :raises SupersetSecurityException: If the user cannot access the resource
         """
 
-        if not self.can_access_datasource(datasource):
-            raise SupersetSecurityException(
-                self.get_datasource_access_error_object(datasource),
-            )
+        from superset import db
+        from superset.connectors.sqla.models import SqlaTable
+        from superset.sql_parse import Table
 
-    def assert_query_context_permission(self, query_context: "QueryContext") -> None:
-        """
-        Assert the the user has permission to access the query context.
+        if database and table or query:
+            if query:
+                database = query.database
 
-        :param query_context: The query context
-        :raises SupersetSecurityException: If the user does not have permission
-        """
+            database = cast("Database", database)
 
-        self.assert_datasource_permission(query_context.datasource)
+            if self.can_access_database(database):
+                return
 
-    def assert_viz_permission(self, viz: "BaseViz") -> None:
-        """
-        Assert the the user has permission to access the visualization.
+            if query:
+                tables = {
+                    Table(table_.table, table_.schema or query.schema)
+                    for table_ in sql_parse.ParsedQuery(query.sql).tables
+                }
+            elif table:
+                tables = {table}
 
-        :param viz: The visualization
-        :raises SupersetSecurityException: If the user does not have permission
-        """
+            inaccessible = set()
+
+            for table_ in tables:
+                schema_perm = self.get_schema_perm(database, schema=table_.schema)
+
+                if not (schema_perm and self.can_access("schema_access", schema_perm)):
+                    datasources = SqlaTable.query_datasources_by_name(
+                        db.session, database, table_.table, schema=table_.schema
```

Comment:
Regarding the `from superset import db`
We could replace it by `self.get_session`